ansible-scylla-node/tasks/upgrade/main.yml

```diff
@@ -0,0 +1,305 @@
+      ansible.builtin.set_fact:
+        scylla_upgrade_version_fix: true
+      when: (upgrade_version_index |int - detected_version_index|int) |abs > 1
+
+    - name: Check if the upgrade can not be straightly applied
+      ansible.builtin.fail:
+        msg: "Version {{ scylla_detected['major_version'] }} can't be upgraded straightly to {{ scylla_upgrade['major_version'] }}."
+      when:
+        - scylla_upgrade_version_fix
+        - not upgrade_autocorrect_version
+
+    - name: Check the next available upgrade for existing Scylla installation
+      vars:
+        distro_name: "{{ ansible_facts['distribution'] |lower }}"
+        distro_version: "{{ ansible_facts['distribution_major_version'] }}"
+        index_delta: "{{ -1 if scylla_upgrade['upgrade'] else 1 }}"
+        detected_version_index: "{{ lookup('ansible.utils.index_of', scylla_support[distro_name][distro_version][scylla_detected['edition']], 'eq', scylla_detected['major_version']) }}"
+        next_available_upgrade_version_index: "{{ detected_version_index |int + index_delta |int }}"
+        next_available_upgrade_version: "{{ scylla_support[distro_name][distro_version][scylla_upgrade['edition']][next_available_upgrade_version_index |int:] |first }}"
+      ansible.builtin.set_fact:
+        scylla_upgrade: {
+          edition: "{{ scylla_upgrade['edition'] }}",
+          edition_friendly_name: "{{ scylla_upgrade['edition_friendly_name'] }}",
+          version: "{{ next_available_upgrade_version }}",
+          major_version: "{{ next_available_upgrade_version.split('.')[0] }}.{{ next_available_upgrade_version.split('.')[1] }}",
+          id: "{{ scylla_upgrade['id'] }}",
+          upgrade: "{{ scylla_upgrade['upgrade'] }}"
+        }
+      when:
+        - scylla_upgrade_version_fix
+        - upgrade_autocorrect_version
+        - next_available_upgrade_version_index |int >= 0
+        - next_available_upgrade_version_index |int < (scylla_support[distro_name][distro_version][scylla_upgrade['edition']] | length)
+
+    - name: Check if the upgrade version was automatically modified
+      ansible.builtin.debug:
+        msg: "Scylla upgrade version was modified to {{ scylla_upgrade['version'] }}"
+      when: scylla_upgrade_version_fix
+  when: scylla_upgrade['edition'] == scylla_detected['edition']
+
+
+# ----- TODO: Scylla cross-edition upgrade validation and downgrade protection -----
+- name: Check Scylla cross-edition upgrade validation and downgrade protection
+  ansible.builtin.fail:
+    msg: "Cross-edition downgrade/upgrade is not implemented yet."
+  when:
+    - scylla_upgrade['edition'] != scylla_detected['edition']
+    - upgrade_major
+    - upgrade_cross_edition
+
+
+# Upgrade
+- name: Trigger pre-upgrade tasks
+  include_tasks: pre_upgrade.yml
+
+- name: Trigger operating system specific upgrade tasks
+  include_tasks: "upgrade_{{ scylla_support[ansible_facts['distribution'] |lower][ansible_facts['distribution_major_version']]['template'] }}.yml"
```

Comment:
I still cannot decide what I like better, above is likely lots of duplication and more maintainance probs
the other way would be single playbook per OS that would handle nitpicks/diffs for appropriate version

@vladzcloudius any take on above approach? (install takes the playbook per OS approach, this is what Eduardo suggested earlier, which gives us more possibilities, but will be harder to maintain and add new versions, so both approaches have their pros and cons)

Reply:
@tarzanek The template is defined for each distro within the matrix, so two distros can share the same template file without any kind if duplicated content. Check Ubuntu, for example.